### ISSUE TITLE
Why is normalization context not set for POST and collection routes?

### ISSUE THREAD

**Issue by tezvi** [api-platform/core#2829]
Hi!

This issue is related to https://github.com/api-platform/core/commit/b1590b8def735cff90db712e9c25c7232ee93765
commited by @teohhanhui, help would be appreciated dude.

This line https://github.com/api-platform/core/blame/master/src/EventListener/ReadListener.php#L66 is preventing building of full request attributes normalization context parameters.

I'm depending on `$request->attributes->set('_api_normalization_context', $normalizationContext);` which is set further bellow.
What is the reason to skip setup of normalization context when POST and collection_operation_name is set?

I'm using custom controllers here and there for handling POST requests and using deserializer group from request attributes when injecting DTO objects in controller action method (using  Request ParamConverter).

Thanks for your input!

**Comment by tezvi** [comment]
Well I'm a bit in a hurry so I managed to fix it on my behalf, though I'm borrowing ApiPlatform internal stuff :\

```
private function getApiPlatformNormalizationContext(Request $request): array
{
    $attributes = RequestAttributesExtractor::extractAttributes($request);

    return $this->contextBuilder->createFromRequest($request, true, $attributes);
}
```

**Comment by teohhanhui** [comment]
Sorry, I don't understand the issue. The `ReadListener` should do nothing if it's a `POST` operation, as there's nothing to be read when you're creating a new resource.

**Comment by tezvi** [comment]
Hi teohhanhui!
I'm using a custom controller for POST operation and I need to know API operation context (ie. serialization / deserialization group) in order to deal with specifics like deserializing and validating non api-platform input etc.
I used to rely on api-platform request attributes which were originally set in ReadListener. Now they are not initialized for POST operations and my code does not work anymore.

So my point is that API operation normalization context attributes should be available in request attributes collection regardless of operation HTTP method, at least to identify that current request is handled by api-platform with specific serialization / deserialization rules.

**Comment by teohhanhui** [comment]
You should use the `RequestAttributesExtractor`, so that's correct. We should perhaps unmark it as `@internal`. @dunglas?

**Comment by tezvi** [comment]
Yup I'm using RequestAttributesExtractor and it would be nice to remove @internal.
Cheers
